src/buf/fixed/pool.rs

```diff
@@ -0,0 +1,341 @@
+    /// an error. Calling `unregister` when no `FixedBufPool` is currently
+    /// registered on this runtime also returns an error.
+    pub fn unregister(&self) -> io::Result<()> {
+        crate::io::unregister_buffers(Rc::clone(&self.inner) as _)
+    }
+
+    /// Returns a buffer of requested capacity from this pool
+    /// that is not currently owned by any other [`FixedBuf`] handle.
+    /// If no such free buffer is available, returns `None`.
+    ///
+    /// The buffer is released to be available again once the
+    /// returned `FixedBuf` handle has been dropped. An I/O operation
+    /// using the buffer takes ownership of it and returns it once completed,
+    /// preventing shared use of the buffer while the operation is in flight.
+    ///
+    /// An application should not rely on any particular order
+    /// in which available buffers are retrieved.
+    pub fn try_next(&self, cap: usize) -> Option<FixedBuf> {
+        let mut inner = self.inner.borrow_mut();
+        inner.try_next(cap).map(|data| {
+            let registry = Rc::clone(&self.inner);
+            // Safety: the validity of buffer data is ensured by
+            // Inner::try_next
+            unsafe { FixedBuf::new(registry, data) }
+        })
+    }
+}
+
+// Internal state shared by FixedBufPool and FixedBuf handles.
+pub(crate) struct Inner {
+    // Pointer to an allocated array of iovec records referencing
+    // the allocated buffers. The number of initialized records is the
+    // same as the length of the states array.
+    raw_bufs: ptr::NonNull<iovec>,
+    // State information on the buffers. Indices in this array correspond to
+    // the indices in the array at raw_bufs.
+    states: Vec<BufState>,
+    // Original capacity of raw_bufs as a Vec.
+    orig_cap: usize,
+    // Table of head indices of the free buffer lists in each size bucket.
+    free_buf_head_by_cap: HashMap<usize, u16>,
+}
+
+// State information of a buffer in the registry,
+enum BufState {
+    // The buffer is not in use.
+    Free {
+        // This field records the length of the initialized part.
+        init_len: usize,
+        // Index of the next buffer of the same capacity in a free buffer list, if any.
+        next: Option<u16>,
+    },
+    // The buffer is checked out.
+    // Its data are logically owned by the FixedBuf handle,
+    // which also keeps track of the length of the initialized part.
+    CheckedOut,
+}
+
+impl Inner {
+    fn new(bufs: impl Iterator<Item = Vec<u8>>) -> Self {
+        let bufs = bufs.take(cmp::min(UIO_MAXIOV as usize, u16::MAX as usize));
+        let (size_hint, _) = bufs.size_hint();
+        let mut iovecs = Vec::with_capacity(size_hint);
+        let mut states = Vec::with_capacity(size_hint);
+        let mut free_buf_head_by_cap = HashMap::new();
+        for (index, mut buf) in bufs.enumerate() {
+            let cap = buf.capacity();
+            let next = free_buf_head_by_cap.insert(cap, index as u16);
```

Comment:
The free list is initially constructed back-to-front, so that for the same capacity bucket, the buffer yielded last will be retrieved first. I did not feel like adding more work or bookkeeping to populate front to back or reverse the list in place, as I don't know of any ill effects this would cause.